### ISSUE TITLE
Mojo principal service crashes with panic sometimes

### ISSUE THREAD

**Issue by aghassemi** [vanadium/issues#943]
I have been getting this panic once in a while. Misha also saw it today.

```
I1124 22:39:00.934858   30858 mojo_blessings.go:104] Obtained OAuth2 token, will exchange for blessings
panic:
runtime error: index out of range

goroutine 17 [running, locked to thread
]:
mojo/public/platform/native_cgo.(*CGoSystem).ReadData(0xf70f9650,
0x16, 0x1, 0x118, 0xf70f9650, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/platform/native_cgo/system_cgo.go:145 +0x104
mojo/public/go/system.(*dataPipeConsumer).ReadData(0xd5f65c88, 0x1, 0x118, 0x0, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/system/data_pipe.go:82 +0x78
main.token2blessings(0xc55fb3b0, 0xd5f76ff0, 0xd6046050, 0x4b, 0xc55ff9e0, 0xd5f659c0, 0x0, 0x0, 0x0, 0x0, ...)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/services/syncbase/syncbased/mojo_blessings.go:159 +0x3ec
main.setBlessings(0xd6020440, 0xc55fb3b0, 0xd5f76ff0, 0x0, 0x0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/services/syncbase/syncbased/mojo_blessings.go:106 +0x174
main.(*delegate).Initialize(0xd5fa7d70, 0xc55fb3b0, 0xd5f76ff0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/services/syncbase/syncbased/mojo_main.go:42 +0x9c
mojo/public/go/application.(*ApplicationImpl).Initialize(0xd5f76ff0, 0xc55fb2e8, 0xd5f657c0, 0xd5f9cde0, 0xd5f6c7e0, 0x53, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/application/application_impl.go:101 +0x16c
mojo/public/interfaces/application/application.(*application_Stub).Accept(0xd5f9cdc0, 0xd5f7ec40, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/out/android_Debug/gen/go/src/mojo/public/interfaces/application/application/application.mojom.go:673 +0x2fc
mojo/public/go/bindings.(*Stub).ServeRequest(0xd5fa1740, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/stub.go:48 +0x98
mojo/public/go/application.Run(0xc55fb1f8, 0xd5fa7d70, 0x4)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/application/application_impl.go:83 +0x228
main.MojoMain(0x4, 0xff900b14)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/services/syncbase/syncbased/mojo_main.go:79 +0xa4

goroutine 20 [syscall, locked to thread]:
mojo/public/platform/native_cgo._Cfunc_MojoWaitMany(0xd601c888, 0xd601c88c, 0x1, 0xffffffff, 0xffffffff, 0xd601c8a0, 0xd601c8a8, 0x0)
    ??:0 +0x38
mojo/public/platform/native_cgo.(*CGoSystem).WaitMany(0xf70f9650, 0xd601c888, 0x1, 0x1, 0xd601c88c, 0x1, 0x1, 0xffffffff, 0xffffffff, 0x0, ...)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/platform/native_cgo/system_cgo.go:190 +0x22c
mojo/public/go/system.(*coreImpl).WaitMany(0xf70f9680, 0xd5f2dac0, 0x1, 0x2, 0xd5f2dad0, 0x1, 0x2, 0xffffffff, 0xffffffff, 0x0, ...)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/system/core.go:100 +0x2b0
mojo/public/go/bindings.(*asyncWaiterWorker).runLoop(0xd5f77040)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/async_waiter.go:177 +0x100
created by mojo/public/go/bindings.newAsyncWaiter
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/async_waiter.go:245 +0x590

goroutine 18 [syscall, locked to thread]:
runtime.goexit()
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/runtime/asm_arm.s:1039 +0x4

goroutine 5 [chan receive]:
github.com/cosnicolaou/llog.(*Log).flushDaemon(0xd5e765a0)
    /usr/local/google/home/aghassemi/git/v23/third_party/go/src/github.com/cosnicolaou/llog/glog.go:1032 +0x64
created by github.com/cosnicolaou/llog.NewLogger
    /usr/local/google/home/aghassemi/git/v23/third_party/go/src/github.com/cosnicolaou/llog/glog.go:546 +0x2c8

goroutine 6 [syscall]:
os/signal.signal_recv(0x0)
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/runtime/sigqueue.go:112 +0x1a8
os/signal.loop()
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/os/signal/signal_unix.go:22 +0x14
created by os/signal.init.1
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/os/signal/signal_unix.go:28 +0x34

goroutine 7 [sleep]:
time.Sleep(0x540be400, 0x2)
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_
android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/runtime/time.go:59 +0x124
v.io/x/ref/lib/stats/sysstats.exportMemStats.func3(0xd5f2da50)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/lib/stats/sysstats/sysstats.go:83 +0x24
created by v.io/x/ref/lib/stats/sysstats.exportMemStats
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/lib/stats/sysstats/sysstats.go:86 +0x25c

goroutine 21 [select]:
mojo/public/go/bindings.(*routerWorker).runLoop(0xd5fa7e60, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:87 +0x308
mojo/public/go/bindings.(*Router).runWorker.func1(0xd5fa7e60, 0xd5fa7e30)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:181 +0x1c
created by mojo/public/go/bindings.(*Router).runWorker
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:204 +0x38

goroutine 23 [chan receive]:
v.io/x/ref/runtime/internal/rt.(*Runtime).addChild.func1(0xd5f7ee40, 0xd5f65960, 0xf6b742f0, 0xf70f9650, 0xd5f65978)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:193 +0x30
created by v.io/x/ref/runtime/internal/rt.(*Runtime).addChild
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:197 +0xe4

goroutine 24 [select, locked to thread]:
runtime.gopark(0xf6e41a48, 0xd5f48f8c, 0xf6cff9e8, 0x6, 0xf61eb718, 0x2)
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/runtime/proc.go:202 +0x15c
runtime.selectgoImpl(0xd5f48f8c, 0x0, 0xc)
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/runtime/select.go:392 +0x1588
runtime.selectgo(0xd5f48f8c)
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/runtime/select.go:212 +0x10
runtime.ensureSigM.func1()
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/runtime/signal1_unix.go:227 +0x448
runtime.goexit()
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/runtime/asm_arm.s:1039 +0x4

goroutine 25 [chan receive]:
v.io/x/ref/runtime/internal/rt.(*Runtime).initSignalHandling.func1(0xd6004f00, 0xd600e7a0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:224 +0x48
created by v.io/x/ref/runtime/internal/rt.(*Runtime).initSignalHandling
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:230 +0x12c

goroutine 26 [chan receive]:
v.io/x/ref/runtime/internal/rt.(*Runtime).addChild.func1(0xd5f7ee40, 0xd5f65960, 0xf6b505e8, 0xd6004f00, 0xd5f659a0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:193 +0x30
created by v.io/x/ref/runtime/internal/rt.(*Runtime).addChild
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:197 +0xe4

goroutine 27 [chan receive]:
v.io/x/ref/runtime/internal/rt.(*Runtime).addChild.func1(0xd5f7ee40, 0xd5f65960, 0xf6cb5eb8, 0xd600f000, 0xd600f880)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:193 +0x30
created by v.io/x/ref/runtime/internal/rt.(*Runtime).addChild
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:197 +0xe4

goroutine 28 [chan receive]:
v.io/x/ref/runtime/internal/rt.(*Runtime).addChild.func1(0xd5f7ee40, 0xd5f65960, 0xf6cd2060, 0xd60059c0, 0xd601c440)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:193 +0x30
created by v.io/x/ref/runtime/internal/rt.(*Runtime).addChild
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:197 +0xe4

goroutine 29 [select]:
v.io/x/ref/runtime/interna
l/flow/manager.NewWithBlessings.func1(0xd600fc60, 0xd6014850)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/flow/manager/manager.go:97 +0x140
created by v.io/x/ref/runtime/internal/flow/manager.NewWithBlessings
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/flow/manager/manager.go:109 +0x808

goroutine 30 [chan receive]:
v.io/x/ref/runtime/internal/rpc.NewXClient.func1(0xd600fc60, 0xd5f77a40)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rpc/xclient.go:89 +0x3c
created by v.io/x/ref/runtime/internal/rpc.NewXClient
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rpc/xclient.go:97 +0x450

goroutine 31 [select]:
v.io/x/ref/runtime/internal/flow/manager.NewWithBlessings.func1(0xd600fe20, 0xd60148c0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/flow/manager/manager.go:97 +0x140
created by v.io/x/ref/runtime/internal/flow/manager.NewWithBlessings
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/flow/manager/manager.go:109 +0x808

goroutine 32 [
chan receive]:
v.io/x/ref/runtime/internal/rpc.NewXClient.func1(0xd600fe20,
0xd5f77a90)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rpc/xclient.go
:89 +0x3c
created by v.io/x/ref/runtime/internal/rpc.NewXClient

    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rpc/xclient.go:97 +
0x450

goroutine 33 [chan receive]:
v.io/x/ref/runtime/internal/rt.(*Runtime).addChild.func1(0xd5f7ee40, 0xd5f65960, 0xf6c8bd40, 0xd600ff60, 0xd601c500
)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:193 +0x30

created by v.io/x/ref/runtime/internal/rt.(*Runtime).addChild
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go
:197 +0xe4

goroutine 34
 [select]:
v.io/x/ref/runtime/internal/flow/manager.NewWithBlessings.func1(0xd6020080
, 0xd6014930)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/flow/manager/manager.go:97 +
0x140
created by v.io/x/ref/runtime/internal/flow/manager.NewWithBlessings
/usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/flow/manager/manager.go:109 +0x808

goroutine 35 [
chan receive]:
v.io/x/ref/runtime/internal/rpc.NewXClient.func1(0xd6020080,
0xd5f77ae0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rpc/xclient.go
:89 +0x3c
created by v.io/x/ref/runtime/internal/rpc.NewXClient
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rpc/xclient.go:
97 +0x450

goroutine
36 [select]:
v.io/x/ref/runtime/internal/flow/manager.NewWithBlessings.func1(0xd6020240, 0xd60149a0
)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/flow/manager/manager.go:97 +
0x140
created by v.io/x/ref/runtime/internal/flow/manager.NewWithBlessings
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/flow/manager/manager.go:109 +0x808

goroutine 37 [chan receive]:
v.io/x/ref/runtime/internal/rpc.NewXClient.func1(0xd6020240, 0xd5f77b30)
/usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rpc/xclient.go:89 +0x3c
created by v.io/x/ref/runtime/internal/rpc.NewXClient
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rpc/xclient.go:97 +0x450

goroutine 38 [chan receive]:
v.io/x/ref/runtime/internal/rt.(*Runtime).addChild.func1(0xd5f7ee40, 0xd5f65960, 0xf6c8bd40, 0xd6020380, 0xd601c5d0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:193 +0x30
created by v.io/x/ref/runtime/internal/rt.(*Runtime).addChild
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/rt/runtime.go:197 +0xe4

goroutine 39 [select]:
v.io/x/ref/lib/pubsub.(*fork).flow(0xd6005f80, 0xd6004ec0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/lib/pubsub/publisher.go:207 +0x2d8
created by v.io/x/ref/lib/pubsub.(*Publisher).CreateStream
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/lib/pubsub/publisher.go:92 +0x3bc

goroutine 40 [syscall]:
syscall.Syscall6(0x124, 0x10, 0xd603a000, 0x1000, 0x0, 0xd6030e4c, 0xd6030e48, 0x0, 0x0, 0x2b)
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/syscall/asm_linux_arm.s:48 +0x8
syscall.recvfrom(0x10, 0xd603a000, 0x1000, 0x1000, 0x0, 0xd6030e4c, 0xd6030e48, 0xc5cb5714, 0x0, 0x0)
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/syscall/zsyscall_linux_arm.go:1331 +0x90
syscall.Recvfrom(0x10, 0xd603a000, 0x1000, 0x1000, 0x0, 0xd603a000, 0x0, 0x0, 0x0, 0x0)
    /usr/local/google/home/aghassemi/git/v23/profiles/go/arm_android_armv7/492a62e945555bbf94a6f9dd6d430f712738c5e0/src/syscall/syscall_unix.go:250 +0x90
v.io/x/lib/netconfig.(*rtnetlinkWatcher).watcher(0xd60205c0)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/lib/netconfig/ipaux_linux.go:325 +0x94
created by v.io/x/lib/netconfig.NewNetConfigWatcher
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/lib/netconfig/ipaux_linux.go:300 +0x384

goroutine 41 [select]:
v.io/x/ref/runtime/internal/lib/roaming.CreateRoamingStream.func1(0xd6005f40, 0xd6020440, 0xd6038580, 0xd6004ec0, 0xc55ffd08, 0xd60205c0, 0xd601c5f0, 0xd5f9cf20, 0x1, 0x1, ...)
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/lib/roaming/roaming.go:48 +0x1190
created by v.io/x/ref/runtime/internal/lib/roaming.CreateRoamingStream
    /usr/local/google/home/aghassemi/git/v23/release/go/src/v.io/x/ref/runtime/internal/lib/roaming/roaming.go:79 +0x22c

goroutine 42 [select]:
mojo/public/go/bindings.(*routerWorker).runLoop(0xd60190e0, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:87 +0x308
mojo/public/go/bindings.(*Router).runWorker.func1(0xd60190e0, 0xd60190b0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:181 +0x1c
created by mojo/public/go/bindings.(*Router).runWorker
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:204 +0x38

goroutine 43 [select]:
mojo/public/go/bindings.(*routerWorker).runLoop(0xd6019230, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:87 +0x308
mojo/public/go/bindings.(*Router).runWorker.func1(0xd6019230, 0xd6019200)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:181 +0x1c
created by mojo/public/go/bindings.(*Router).runWorker
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:204 +0x38

goroutine 51 [select]:
mojo/public/go/bindings.(*routerWorker).runLoop(0xd60164b0, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:87 +0x308
mojo/public/go/bindings.(*Router).runWorker.func1(0xd60164b0, 0xd6016480)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:181 +0x1c
created by mojo/public/go/bindings.(*Router).runWorker
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:204 +0x38

goroutine 52 [select]:
mojo/public/go/bindings.(*routerWorker).runLoop(0xd60165d0, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:87 +0x308
mojo/public/go/bindings.(*Router).runWorker.func1(0xd60165d0, 0xd60165a0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:181 +0x1c
created by mojo/public/go/bindings.(*Router).runWorker
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:204 +0x38

goroutine 53 [select]:
mojo/public/go/bindings.(*routerWorker).runLoop(0xd6016690, 0x0, 0x0)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindi
ngs/router.go:87 +0x308
mojo/public/go/bindings.(*Router).runWorker.func1(0xd6016690, 0xd6016660)
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:181 +0x1c
created by mojo/public/go/bindings.(*Router).runWorker
    /usr/local/google/home/aghassemi/mojo/src/mojo/public/go/bindings/router.go:204 +0x38
```

**Comment by asimshankar** [comment]
This is a duplicate of domokit/modular#351 and I have sent out a change to fix this https://codereview.chromium.org/1463643002/ but am waiting on the review

**Comment by aghassemi** [comment]
Sweet. Thanks Asim. Could you maybe ping the reviewers? Looks like you sent out the fix a while ago!

**Comment by afandria** [comment]
Thanks Asim, I'm running into this a lot too.
